src/core_plugins/kibana/public/discover/controllers/discover.js

```diff
@@ -348,56 +348,57 @@ function discoverController(
+          };
+
+          function pick(rows, oldRows, fetchStatus) {
+            // initial state, pretend we are loading
+            if (rows == null && oldRows == null) return status.LOADING;
+
+            const rowsEmpty = _.isEmpty(rows);
+            // An undefined fetchStatus means the requests are still being
+            // prepared to be sent. When all requests are completed,
+            // fetchStatus is set to null, so it's important that we
+            // specifically check for undefined to determine a loading status.
+            const preparingForFetch = _.isUndefined(fetchStatus);
+            if (preparingForFetch) return status.LOADING;
+            else if (rowsEmpty && fetchStatus) return status.LOADING;
+            else if (!rowsEmpty) return status.READY;
+            else return status.NO_RESULTS;
+          }
 
-            function pick(rows, oldRows, fetchStatus) {
-              // initial state, pretend we are loading
-              if (rows == null && oldRows == null) return status.LOADING;
-
-              const rowsEmpty = _.isEmpty(rows);
-              // An undefined fetchStatus means the requests are still being
-              // prepared to be sent. When all requests are completed,
-              // fetchStatus is set to null, so it's important that we
-              // specifically check for undefined to determine a loading status.
-              const preparingForFetch = _.isUndefined(fetchStatus);
-              if (preparingForFetch) return status.LOADING;
-              else if (rowsEmpty && fetchStatus) return status.LOADING;
-              else if (!rowsEmpty) return status.READY;
-              else return status.NO_RESULTS;
-            }
+          return function () {
+            const current = {
+              rows: $scope.rows,
+              fetchStatus: $scope.fetchStatus
+            };
 
-            return function () {
-              const current = {
-                rows: $scope.rows,
-                fetchStatus: $scope.fetchStatus
-              };
+            $scope.resultState = pick(
+              current.rows,
+              prev.rows,
+              current.fetchStatus,
+              prev.fetchStatus
+            );
 
-              $scope.resultState = pick(
-                current.rows,
-                prev.rows,
-                current.fetchStatus,
-                prev.fetchStatus
-              );
+            prev = current;
+          };
+        }()));
 
-              prev = current;
-            };
-          }()));
+        // function initForTime() {
+        //   return setupVisualization().then($scope.updateTime);
+        // }
 
-        function initForTime() {
-          return setupVisualization().then($scope.updateTime);
+        if ($scope.opts.timefield) {
+          setupVisualization();
+          $scope.updateTime();
         }
 
-        return Promise.resolve($scope.opts.timefield && initForTime())
```

Comment:
That promise was the actual returning value of the above

```js
$scope.updateDataSource()
      .then(...)
```

Since that value itself is never returned nor acted upon, we can simply remove returning anything from this method.

Also since we don't need to wait for `setupVisualization()` anymore (see comment below), we can easily replace this logic by a simple `if ($scope.opts.timefield)` and executing the previous resolve function synchronous before this method ends.